src/Rewards.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+import {NODL} from "./NODL.sol";
+import {AccessControl} from "openzeppelin-contracts/contracts/access/AccessControl.sol";
+import {EIP712} from "openzeppelin-contracts/contracts/utils/cryptography/EIP712.sol";
+import {ECDSA} from "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title Nodle DePIN Rewards
+ * @dev This contract allows an authorized oracle to issue off-chain signed rewards to recipients.
+ * This contract must have the MINTER_ROLE in the NODL token contract.
+ */
+contract Rewards is AccessControl, EIP712 {
+    /**
+     * @dev Role required to set the reward quota.
+     */
+    bytes32 public constant QUOTA_SETTER_ROLE = keccak256("QUOTA_SETTER_ROLE");
+
+    /**
+     * @dev The signing domain used for generating signatures.
+     */
+    string public constant SIGNING_DOMAIN = "rewards.depin.nodle";
+
+    /**
+     * @dev The version of the signature scheme used.
+     */
+    string public constant SIGNATURE_VERSION = "1";
+
+    /**
+     * @dev This constant defines the reward type.
+     * This should be kept consistent with the Reward struct.
+     */
+    bytes public constant REWARD_TYPE = "Reward(address recipient,uint256 amount,uint256 counter)";
+
+    /**
+     * @dev Reference to the NODL token contract.
+     */
+    NODL public nodlToken;
+
+    /**
+     * @dev Maximum amount of rewards that can be distributed in a period.
+     */
+    uint256 public rewardQuota;
+
+    /**
+     * @dev Duration of each reward period.
+     */
+    uint256 public rewardPeriod;
+
+    /**
+     * @dev Timestamp indicating when the reward quota is due to be renewed.
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward counters for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardCounters;
```

Comment:
can we rename from counter to nonce? It is effectively a nonce

Reply:
```suggestion
    mapping(address => uint256) public nonces;
```

Reply:
The reason I went for `counter` than `nonce` was to distinguish it from a transaction nonce. It's an app level nonce I didn't want to confuse anyone in the future.

Reply:
I have renamed it to `sequence`. Is it good?